libnopegl/src/glsl/helper_noise.glsl

```diff
@@ -0,0 +1,179 @@
+}
+
+float random(uvec3 x)
+{
+    return u32tof32(hash(x));
+}
+
+float random_grad(uvec2 x, vec2 f)
+{
+    float r = u32tof32(hash(x)) * 2.0 * ngli_pi;
+    return dot(vec2(cos(r), sin(r)), f);
+}
+
+float random_grad(uvec3 x, vec3 f)
+{
+    uint hash = hash(x) & 15U;
+    float u = hash < 8U ? f.x : f.y;
+    float vt = ((hash == 12U) || (hash == 14U)) ? f.x : f.z;
+    float v = hash < 4U ? f.y : vt;
+    return ((hash & 1U) != 0U ? -u : u) + ((hash & 2U) != 0U ? -v : v);
+}
+
+float noise_perlin(vec2 t, uint seed)
+{
+    vec2 i = floor(t);
+    vec2 f = fract(t);
+    uvec2 x = uvec2(i) + seed;
+
+    float y0 = random_grad(x + uvec2(0, 0), f - vec2(0, 0));
+    float y1 = random_grad(x + uvec2(1, 0), f - vec2(1, 0));
+    float y2 = random_grad(x + uvec2(0, 1), f - vec2(0, 1));
+    float y3 = random_grad(x + uvec2(1, 1), f - vec2(1, 1));
+
+    vec2 a = f * f * f * (f * (f * 6.0 - 15.0) + 10.0);
+    return(mix(mix(y0, y1, a.x), mix(y2, y3, a.x), a.y));
+}
+
+float noise_perlin(vec3 t, uint seed)
+{
+    vec3 i = floor(t);
+    vec3 f = fract(t);
+    uvec3 x = uvec3(i) + seed;
+
+    float y0 = random_grad(x + uvec3(0, 0, 0), f - vec3(0, 0, 0));
+    float y1 = random_grad(x + uvec3(1, 0, 0), f - vec3(1, 0, 0));
+    float y2 = random_grad(x + uvec3(0, 1, 0), f - vec3(0, 1, 0));
+    float y3 = random_grad(x + uvec3(1, 1, 0), f - vec3(1, 1, 0));
+    float y4 = random_grad(x + uvec3(0, 0, 1), f - vec3(0, 0, 1));
+    float y5 = random_grad(x + uvec3(1, 0, 1), f - vec3(1, 0, 1));
+    float y6 = random_grad(x + uvec3(0, 1, 1), f - vec3(0, 1, 1));
+    float y7 = random_grad(x + uvec3(1, 1, 1), f - vec3(1, 1, 1));
+
+    vec3 a = f * f * f * (f * (f * 6.0 - 15.0) + 10.0);
+    return mix(
+               mix(mix(y0, y1, a.x),
+                   mix(y2, y3, a.x),
+                   a.y),
+               mix(mix(y4, y5, a.x),
+                   mix(y6, y7, a.x),
+                   a.y),
+               a.z);
+}
+
+/* Adapted from https://en.wikipedia.org/wiki/Bicubic_interpolation */
+float bicubic(float a, float b, float c, float d, float t)
+{
+    mat4 m = 0.5 * mat4(
+        0.,  2.,  0.,  0.,
+       -1.,  0.,  1.,  0.,
+        2., -5.,  4., -1.,
+       -1.,  3., -3.,  1.
+    );
+    return dot(m * vec4(1, t, t*t, t*t*t), vec4(a, b, c, d));
+}
+
+float noise_block(vec3 t, uint seed)
+{
+    vec3 i = floor(t);
+    vec3 f = fract(t);
+    uvec3 x = uvec3(i) + seed;
+
+    float freq_x = random(uvec3(x.xy, seed));
+    float freq_y = random(uvec3(x.xy, seed << 16U));
+    float evolution = t.z * freq_x + freq_y;
+    uint e_i = uint(floor(evolution));
+    float e_f = fract(evolution);
+
+    float y0 = random(uvec3(x.xy, e_i - 1U));
+    float y1 = random(uvec3(x.xy, e_i + 0U));
+    float y2 = random(uvec3(x.xy, e_i + 1U));
+    float y3 = random(uvec3(x.xy, e_i + 2U));
+
+    float res = bicubic(y0, y1, y2, y3, e_f);
+
+    /* Scale noise from the [0;1] range to [-1;1] */
+    return res * 2.0 - 1.0;
```

Comment:
why is this done only for blocky?

Reply:
Because random returns value in range [0; 1]